x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/SequenceRuntime.java

```diff
@@ -26,105 +31,103 @@
+            listener.onResponse(sequencePayload());
             return;
         }
+
         // else continue finding matches
         Criterion currentCriterion = criteria.get(stage);
-        // narrow by the previous stage timestamp marker
-        currentCriterion.fromMarkers(stateMachine.getMarkers(stage - 1));
+        if (stage > 0) {
+            // FIXME: revisit this during pagination since the second criterion need to be limited to the range of the first one
+            // narrow by the previous stage timestamp marker
+
+            Criterion previous = criteria.get(stage - 1);
+            // if DESC, flip the markers (the stop becomes the start due to the reverse order), otherwise keep it accordingly
+            Object[] marker = descending && stage == 1 ? previous.stopMarker() : previous.startMarker();
+            currentCriterion.useMarker(marker);
+        }
         
-        queryClient.query(currentCriterion.searchSource(), wrap(payload -> {
-            findMatches(stage, payload);
-            inspectStage(stage + 1, resultsListener);
-        }, resultsListener::onFailure));
+        log.info("Querying stage {}", stage);
+        queryClient.query(currentCriterion, wrap(payload -> {
+            List<SearchHit> hits = payload.values();
+
+            // nothing matches the query -> bail out
+            // FIXME: needs to be changed when doing pagination
+            if (hits.isEmpty()) {
+                listener.onResponse(sequencePayload());
+                return;
+            }
+
+            findMatches(stage, hits);
+            queryStage(stage + 1, listener);
+        }, listener::onFailure));
     }
 
-    private void findMatches(int currentStage, Payload<SearchHit> payload) {
-        Criterion currentCriterion = criteria.get(currentStage);
-        List<SearchHit> hits = payload.values();
-        
+    // hits are guaranteed to be non-empty
+    private void findMatches(int currentStage, List<SearchHit> hits) {
+        // update criterion
+        Criterion criterion = criteria.get(currentStage);
+        criterion.startMarker(hits.get(0));
+        criterion.stopMarker(hits.get(hits.size() - 1));
+
         // break the results per key
-        for (SearchHit hit : hits) {
-            KeyAndOrdinal ko = findKey(hit, currentCriterion);
-            stateMachine.match(currentStage, ko.key, ko.timestamp, ko.tiebreaker, hit);
+        // when dealing with descending order, queries outside the base are ASC (search_before)
+        // so look at the data in reverse (that is DESC)
+        Iterator<SearchHit> iterator = descending ? new ReversedIterator<>(hits) : hits.iterator();
+
+        for (; iterator.hasNext();) {
+            SearchHit hit = iterator.next();
+
+            // early skip in case of reaching the limit
+            // check the last stage to avoid calling the state machine in other stages
+            if (isLast(currentStage) && stateMachine.reachedLimit()) {
```

Comment:
I've mentioned in the comment - checking the reachedLimit all the time does make sense but at the same time it will only increase during the last stage. Hence why checking it before is not needed.
Nevertheless I can see how this can be confusing plus I've improved the stateMachine class to have a bool and avoid the check on each method call.